enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+
+## Motivation
+
+The primary purpose of this enhancement is to define how OpenShift can protect existing services introduced by OLM Managed Operators when performing a Minor OpenShift upgrade.
+
+### Goals
+
+- Prevent Minor OpenShift Upgrades  when an installed operator declares that it does not support the next ocp minor version
+- Warn the Cluster Admin when an installed operator does not explicitly declare support for the next ocp minor version
+
+### Non-Goals
+
+- Ensure that services will not be disrupted during a cluster upgrade.
+- Guarantee that all installed operator will continue to run after a cluster upgrade.
+- Prevent Patch OpenShift Upgrades.
+
+## Proposal
+
+The [ClusterVersionOperator](https://github.com/openshift/cluster-version-operator) provides OLM with the means to prevent Minor Semantic Version cluster upgrades by setting the  `operator-lifecycle-manager` [ClusterOperator's](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md) [upgradeable condition](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource) to `False`.
+
+OLM currently reports that it is upgradeable as soon as the OLM Deployment is successfully rolled out. This existing logic should be updated to monitor the collection of OLM Managed Operators on cluster and update the `upgradeable condition` to reflect if any of these operators indicate that they will not run on the next Minor OpenShift Version.
+
+The bulk of this enhancement focuses on defining the steps that OLM will take to determine upgrade safety based on the collection of installed operators.
+
+### Allowing Operator Authors to define supported OpenShift Versions
+
+Many Operators shipped by Red Hat are rigorously tested on a specific set of OpenShift Versions. The teams responsible for this testing know exactly which versions of OpenShift that their Operators can run on. OLM should allow Operator Authors to specify the maximum Major and Minor version of OpenShift that their operator may run on.
+
+Given that this is an OpenShift specific feature, OLM would rather not expand the Spec of the [ClusterServiceVersion (CSV)](https://olm.operatorframework.io/docs/concepts/crds/clusterserviceversion/), used on vanilla Kubernetes clusters, to include a `maxOpenShiftVersion` field. Instead, OLM will rely on the presence of an annotation on the CSVs to specify the max OpenShift Version that the operator may run on. This annotation can be added in one of two ways:
+
+1. By defining the Max OpenShift Version as a property of the bundle using the [Declarative Index Config](https://github.com/operator-framework/enhancements/blob/1cf4a0363d918e810f638036539388622265d466/enhancements/declarative-index-config.md) (recommended approach)
+2. By adding the annotation to the CSV Directly
```

Comment:
Is this the only option for #357, where the compatibility with the next OpenShift version is less about the OLM-managed operator and more about that OLM-child's operands (in the case of #357, the kernel modules being installed)?  I guess the special-resource operator would check version compat of the kernel modules it managed vs. some compiled-into-the-SRO knowledge of the kernel version that was likely coming with the next OCP minor, and then manage the annotation in its own CSV to pass the word on to OLM?

Reply:
> I guess the special-resource operator would check version compat of the kernel modules it managed vs. some compiled-into-the-SRO knowledge of the kernel version that was likely coming with the next OCP minor, and then manage the annotation in its own CSV to pass the word on to OLM?

Not sure if you're suggesting the operator updates its own CSV at runtime to change its maxVersion.  If so, i don't think we had that in mind (not even sure if you can do that, or if OLM will stomp your attempted changes).

if you're just saying the operator team would set the annotation on their CSV prior to shipping the operator bundle, then yes, that's the expectation.

Reply:
> if you're just saying the operator team would set the annotation on their CSV prior to shipping the operator bundle, then yes, that's the expectation.

Or the annotation could be set using the declarative index config.